### PR TITLE
Sensors

### PR DESCRIPTION
**Pull request by doadin**
Needs Review

Based on:
https://github.com/Christopher83/samsung-kernel-msm7x30/commit/d8ff5ba3ca6aa552566cfa99794300679083ba7b#diff-7

### CODE REVIEW

**Comment by CastagnaIT** [comment]
attention if you change parameters with my commit is required to update all the remain parts:
device and vendor/deamons and ramdisk

otherwise it will not work well,
i do not recommend to do a partial porting

**Comment by arco** [comment]
Yeah I was gonna say that too. Not sure I wanna mess with this stuff right now, especially when I don't have much free time fixing stuff it might break. I know that at least the sensor libs and in one specific place in the kernel input drivers, it's looking for the "accelerometer" name, so changing that might break things.

**Comment by CastagnaIT** [comment]
Hi arco,i understand
if someone want to port from ariesve this is the full commits:

Device part:
https://github.com/AriesVE-DevCon-TEAM/android_device_samsung_ariesve/commit/5a7df62d9abc8ff32502954d43e893e97a8e7be1

Kernel part:
https://github.com/AriesVE-DevCon-TEAM/samsung-kernel-msm7x30/commit/63271587aa4e45c146be982c890dcdc8796ebfa2

Vendor part:
https://github.com/AriesVE-DevCon-TEAM/android_vendor_samsung_ariesve/commit/e248cb20e7e71bccb834dc6581e13881ec1ef88c

**Comment by doadin** [comment]
Yea, i know about the other parts too. If this looks good and i didnt mess anything up i was gonna move forward with the other changes.

**Comment by doadin** [comment]
I was planing oon makeing forks for both ancora and ancora_tmo and test on the tmo side before makeing a pull request for ancora.

**Comment by CastagnaIT** [comment]
ok good work :)

**Comment by doadin** [comment]
I know alot of devs watch these repos and what not and was just sticking this all out there incase anyone see anything i may have missed or didnt change properly as i go and then im sure before closeing this arco would want to test for himself. As i put in the comment at least for this section here it hold most of the kernel side changes for now at least for ancora_tmo. Ill add ancora later. In addition to makeing a fork/seperate request for the device/vendor counterparts.

Which all of this is in thanks to you CastagnaIT  and Christopher and the whole dev team and what not.

I know i havent really added in your names to commits but we can add that later one all the changes are figured out.

**Comment by doadin** [comment]
Most of the sensor drivers appear to be the same. The Areiseve seems to have just had newer driver in addition to the changes made so this part was relativly easy for me.(Im sure you guys put in alot of time, thanks) So ill add in the rest in the next few days. Thanks for links to the other changes. Itll make it easyer to see all the parts so hopefully its be fairly quick.

**Comment by doadin** [comment]
One thing im unsure of is the light sensor, i dont believe the ancora/ancora_tmo has one? If thats the case do i need to take that part out of the device side? Or will it just be kinda like bloat and just not do anything?

Also since we dont use anything like AriesParts is there anything like that that needs to be changed device side for us? Im guessing no but not 100% sure. 

I went ahead and made the other device/vendor changes but havent made a pull request yet.

Also just wana say i understand too arco im like just trying to keep up with all the nice things the AriesVE-DevCon-TEAM is makeing for the ariesve.

**Comment by doadin** [comment]
Besides those little questions. The only thing that i dont think lines up is the ak8975 device for ancora. Im unsure if that needs any changeing or not. Like i was saying i got most of this from the changes  CastagnaIT made just trying to adapt them which is close except this one last part, i think.

**Comment by arco** [comment]
drivers/sensors_ancora/bma.h:166:31: error: 'BMA_MAGCALIB_SHAPE_NUM' undeclared here (not in a function)

**Comment by doadin** [comment]
I dont have much time tonight. Looking real quick though i dont see that anywhere else in new or old ancora or Ariesve except in the same place. Ill look later. Some of the driver are combined with ancora wheres they are seperate in Ariesve so maybe its in another file. Ill look through the includes on everything.

**Comment by doadin** [comment]
Found it lol

**Comment by doadin** [comment]
opps lol

**Comment by CastagnaIT** [comment]
acutally you need to make some changes, (the light sensor ancora dont have)
for AK8975 compass sensor you can try to find the source code from galaxy r
and adapt it for your device.
The ariespart is not really needed, however is easy to add it,
but you can not copy it directly from ariesve because we have proprietary options that ancora not have...

**Comment by doadin** [comment]
Ok, thanks for the info.

**Comment by doadin** [comment]
Not sure about lines 65-190 of ak8975.c Arieseve has another include which is tied to this. Maybe take part of that include that we need and add it as a new file and new include.

Also unsure about the missing input_report_abs/#define REL_STATUS  and all of that. Although i think that was just to make it similar to what new kernel drivers use and is not needed to change over to the opensource libsensor? Although would probably be nice to change over to in the future, if thats the case.

(partially note to self)

**Comment by doadin** [comment]
I think those 2 should about finish it up although im not sure the second one will be ok. We need part of it for the 65-190 lines of ak8975.c but not sure if we need everything in there or if it will break compileing.(Or if there needs to be more?) Ill have some time to test that later this week.

**Comment by CastagnaIT** [comment]
I hope you're not doing a simple copy paste without knowing what you're doing..
if you need to make a port, it is mandatory to fully examine the new and the old drivers to understand the differences,
i would give you a help but i don't have time...sorry

**Comment by doadin** [comment]
Yes and no. With everything except ak8975 you guys did all the work so simple copy paste was all that was needed. ak8975 is proving to be a litttle more complex. I am still taking parts from but changing it up as i find the differences. Im just not as good at this so i might have alot of changes in here and it might get a little messy lol.

**Comment by arco** [comment]
drivers/sensors_ancora/bma222.c:1771:5: error: implicit declaration of function 'bma222_set_calibration' [-Werror=implicit-function-declaration]
drivers/sensors_ancora/bma222.c: In function 'bma_get_calibration':
drivers/sensors_ancora/bma222.c:1790:2: error: implicit declaration of function 'bma222_get_calibration' [-Werror=implicit-function-declaration]

**Comment by doadin** [comment]
^ 
Is how it was before so that should be ok now? 

Was that the only error?

**Comment by doadin** [comment]
I know there was like always alot of warnings but the commit before this should have cleared up the other errors i think.

So this should be like final touches. If that fixes that error next ill try a while rom build. which btw if you look on my github page i have a fork with device/vendor tree if you want to try it.(If thats the last error and if this fixes it lol lot of ifs)

I have only had one chance to try compiling and the only time i have to make changes is while im working on other work siting here at the computer.

All the code is all jumbled up. If this all works next maybe ill try and clean it up lol although after this prolly wont need to touch it ever again lol.

**Comment by doadin** [comment]
It all compiles now, so maybe its all done? I guess im going to make a rom build and see later.

**Comment by arco** [comment]
drivers/built-in.o: In function `geomagnetic_probe':
ak8975.c:(.text+0x1e5d98): undefined reference to`sensors_register'
make: **\* [.tmp_vmlinux1] Error 1

You know that this has to work on 3 different devices? The ancora, ancora_tmo and apache. There's a lot of checks in the original drivers that checks which device it's compiled for and adjusts various parts accordingly.

**Comment by doadin** [comment]
That is odd, it compiled just fine for me last night.

And yes i know it needs to work for several devices. If you look at all that was changed everything in the drivers them self should be the same, no device specific changes where made. The only additional changes that are needed are in the board-_devicename_.c, as far as the kernel goes. In which id gladly change that when this is finalized. And i have already made device/vendor trees for ancora and ancora_tmo and if theres anyone that needs help id gladely make one for apache too. device side changes are easy though. I think the hardest part would be getting people to know about the change.

**Comment by doadin** [comment]
Ah i think i see now what the difference might be, did you not pull the changes to ak8975-reg.h? It has 

int sensors_register(struct device *dev, void \* drvdata,
    struct device_attribute *attributes[], char *name);

Now for that new code.

**Comment by arco** [comment]
I pulled everything that was in the pull request. :)
I'll try again later.

**Comment by doadin** [comment]
If you are still haveing problems i guess you can try removeing lines 637-642. That shouldnt break anything and should compile if that was the only error.

**Comment by doadin** [comment]
I guess i spoke too soon right before makeing the zImage it get6s that error. I guess i assumed once it compiles it earlyer on that it was fine. Im testing again but if you remove the above mentioned lines i think it will work.

Edit: Confirmed does compile 100% no errors now after removeing that line.

**Comment by doadin** [comment]
It appears it doesnt want to boot. Not sure why but i dont know what to do so might just have to scrap this whole thing unless someone else can find the issue.

The only thing i didnt add is on ariesvs side they have, libsensor_yamaha_test.so. Not sure if thats needed or not. Or if it would cause it not to boot.

**Comment by doadin** [comment]
I took down my last message because i thought i had flashed the wrong kernel but i just double checked and it does in fact boot now. This test was with only the kernel changed not rom so i guess its still possible it could not boot with the rom changes but for now at least i know there are no kernel boot issues. Will update later after my rom build finishes.

**Comment by doadin** [comment]
Ok, so far im at with just kernel changes only proximity sensor work but accelaromiter and compass do not. At this point if i try to just replace the sensor lib with the new one it then bootloops.(probably because it need the rom changes)

Also in logcat with just kernel changes it says something along the lines of sensor enable 1 device doesnt exist.

Im guessing all the current issues are rom related. will build tonight and test.

**Comment by doadin** [comment]
Ok, one more progress update. I got it to boot with both kernel and rom changes however a few things are happening.
1. kernel dmesg spaming <4>[  245.472076] [diony] bma_acc_work_func.
2. The compass doesnt work.
3. Acceleromiter seems to be giving opposite info. for example i got a app to sensors and for the acceleromiter test theres a ball on the screen and you move it by tilting the phone. with the changes you have to move the phone in the opposite direction for the ball to move to the left side of the screen you tilt the phone right wheres without the changes you would tilt it left.

Proximity does work.

Maybe the wrong values have something to do with the compiler warrnings:
CC drivers/sensors_ancora/bma_accel_driver.o
In file included from drivers/sensors_ancora/bma_accel_driver.c:37:0:
drivers/sensors_ancora/bma222.c: In function ‘bma222_set_calibration’:
drivers/sensors_ancora/bma222.c:849:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
drivers/sensors_ancora/bma222.c: In function ‘bma_set_calibration’:
drivers/sensors_ancora/bma222.c:1927:5: warning: passing argument 1 of ‘bma222_set_calibration’ from incompatible pointer type [enabled by default]
drivers/sensors_ancora/bma222.c:844:12: note: expected ‘signed char _’ but argument is of type ‘signed char *_’
drivers/sensors_ancora/bma222.c: In function ‘bma_acc_driver_init’:
drivers/sensors_ancora/bma222.c:1993:24: warning: assignment from incompatible pointer type [enabled by default]
drivers/sensors_ancora/bma_accel_driver.c: In function ‘bma_acc_i2c_read’:
drivers/sensors_ancora/bma_accel_driver.c:234:17: warning: passing argument 1 of ‘printk’ from incompatible pointer type [enabled by default]
In file included from include/linux/kernel.h:22:0,
                 from include/linux/delay.h:10,
                 from drivers/sensors_ancora/bma_accel_driver.c:19:
include/linux/printk.h:101:5: note: expected ‘const char _’ but argument is of type ‘struct device *’
drivers/sensors_ancora/bma_accel_driver.c: In function ‘bma_acc_fast_calibration’:
drivers/sensors_ancora/bma_accel_driver.c:416:5: warning: passing argument 1 of ‘driver->set_calibration’ from incompatible pointer type [enabled by default]
drivers/sensors_ancora/bma_accel_driver.c:416:5: note: expected ‘unsigned char *’ but argument is of type ‘unsigned char *_’
drivers/sensors_ancora/bma_accel_driver.c: In function ‘bma_acc_enable_store’:
drivers/sensors_ancora/bma_accel_driver.c:496:4: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘long unsigned int’ [-Wformat]

**Comment by doadin** [comment]
I actually think i may have figured out the issues. Im like completely burnt out on make rom builds over and over again though so if someone else wants to try, go for it i think it might be done now with ^that change and the new rom side changes.

**Comment by doadin** [comment]
I guess im done. I for the life of me cant figure out whats going wrong with accel. As this commit is now The accel reports oppsoite values and compass doesnt work. proximity is the only one that works. Gonna close this commit.

I wish i could have figured it out because the accel run alot smoother and with no stutter but oh well. but hey i tried.

**Comment by arco** [comment]
Appreciate all the work doadin. But yeah, getting this stuff to work correctly isn't easy, especially when it also has to be in sync with user space HAL that we don't have sources for.

**Comment by doadin** [comment]
Well i have a feeling the accel could probably be calibrated with an app like the ariesparts like  CastagnaIT said could be ported. And the proximity does work so its only the compass really, but im just kinda worn out, im a one man team here(no ancora_tmo devs help) lol. Maybe ill revisit it later.

**Comment by doadin** [comment]
I found source code from the company who makes the AK8975(compass sensor). Apparently they have released kernel and libsensor source. However im not sure if any/all of it can be used or not?!

Obviously if we can switch things like AKM_I2C_NAME would need to be changed but idk.

It can be seen here: https://github.com/akm-multisensor/AK8975_FS

**Comment by hey-o** [comment]
Google has something, too.
https://android.googlesource.com/platform/hardware/akm/+/master

**Comment by arco** [comment]
OEM's usually heavily modifies the sources so it works with how the sensor is wired in the device. Even if it's the same sensor used in many devices, almost all implementations are different. Like some uses a helper lib (like we have with libakm.so), and some don't. It was way easier a few years ago when I was working with the HTC Tattoo and Wildfire, because the implementation was pretty much the same on a lot of HTC devices. So we could use the same opensource HAL with just minor modifications. Today it's a lot more complicated, and on new devices no one really bothers anymore with opensource HAL's.

**Comment by doadin** [comment]
Yea the only difference i see that would be device specific is how the irq works. In which its the same as open source code from other samsung device which didnt work. I think it would probably be easyer to just use the originial akm libsensor part i had originally. The only problem is i pulled it from a cyanogenmod review that was merged but for w.e. reason it doesnt compile for me there must be something im missing but i dont know what it is yet.

**Comment by doadin** [comment]
I found the one part it was just a simple missing define. I think it only fixed part of the problem though so ill have to figure out the rest later. If anyone wants to look i found it at the review here: http://review.cyanogenmod.org/#/c/49811/2 .

**Comment by hey-o** [comment]
Any update? Is it abandoned? (Interesting, might be useful: http://redmine.replicant.us/boards/21/topics/693)

**Comment by doadin** [comment]
Not necessarily abandoned just havent had much time. And yes i saw that post before and i had spoken with ralisi but didnt get a whole lot from him. That might work but i havent tried it. If you pull the libsensor folder from my ancora_tmo repo and replace just the akm parts that would be a good point for starting testing. The akm part is the last one. The kernel side works but i didnt get a working userspace. It maybe because of the akm fuse part. Looking at the samsung .so with ida i didnt see anything about it and the other opensource versions ive seen didnt use it except the one from the akm company. 

Long story short i have not tried that version of the userspace driver and the userspace driver for akm is the last thing needed to move all the sensors opensource. If you try it let me know how it works.

Edit: Also id be more then happy to test and provide logs but other than that for the time being i dont have much time for anything more.

**Comment by ralisi** [comment]
About that linked thread from replicant:

Sorry, I didn't find the necessary time to continue working on the sensors.

To be honest, I am not even sure if the linked code really works, or if I just failed to remove correctly the proprietary .so

**Comment by doadin** [comment]
Apparently i got rid of this "patch-1" branch in my repo. If we could keep this open just in case(dont wana lose this progress) that would be nice. However since i couldnt make a change to a branch that doesnt exist im going to make a new pull request with changes i made today.
